File: src/backend/api/v1/game.py
"""
Game API v1 routes.

Provides endpoints for game actions, state management, and dice operations.
"""

import uuid
from typing import Any

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field

from src.backend.agents.npc import NPCAgent

router = APIRouter(prefix="/api/v1", tags=["game"])


class NewGameRequest(BaseModel):
    """Request model for starting a new game."""

    world_pack_id: str = Field(default="demo_pack", description="World pack to load")
    player_name: str = Field(
        default="玩家", description="Player (user) name - distinct from character name"
    )
    preset_character_id: str | None = Field(
        default=None, description="ID of preset character to use (from world pack)"
    )


class NewGameResponse(BaseModel):
    """Response model for new game creation."""

    session_id: str
    player: dict[str, Any]
    game_state: dict[str, Any]
    world_info: dict[str, Any]
    starting_scene: dict[str, Any]
    message: str


@router.post("/game/new", response_model=NewGameResponse)
async def start_new_game(request: NewGameRequest):
    """
    Start a new game session.

    Creates a new game session with the specified world pack and player info.
    Loads the world pack and sets up the starting location with full scene info.

    Args:
        request: NewGameRequest containing world_pack_id and player info

    Returns:
        NewGameResponse with session_id, player data, initial game state,
        world info, and starting scene description
    """
    from src.backend.main import get_world_pack_loader, gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    world_pack_loader = get_world_pack_loader()
    if world_pack_loader is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="World pack loader not initialized",
        )

    try:
        # Generate a new session ID
        session_id = str(uuid.uuid4())

        # Load the world pack
        try:
            world_pack = world_pack_loader.load(request.world_pack_id)
        except FileNotFoundError as exc:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"World pack not found: {request.world_pack_id}. "
                f"Available packs: {world_pack_loader.list_available()}",
            ) from exc
        except ValueError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid world pack: {exc}",
            ) from exc

        # Find starting location (look for "starting_area" tag)
        starting_location_id = None
        starting_location = None
        for loc_id, loc in world_pack.locations.items():
            if "starting_area" in loc.tags:
                starting_location_id = loc_id
                starting_location = loc
                break

        # Fallback to first location if no starting_area tag found
        if starting_location_id is None and world_pack.locations:
            starting_location_id = next(iter(world_pack.locations.keys()))
            starting_location = world_pack.locations[starting_location_id]

        if starting_location is None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="World pack has no locations defined",
            )

        # Get NPCs at starting location
        active_npc_ids = starting_location.present_npc_ids or []

        # Register NPC Agents for NPCs in the starting location
        # First, remove any existing NPC agents from sub_agents
        npc_keys_to_remove = [k for k in gm_agent.sub_agents if k.startswith("npc_")]
        for key in npc_keys_to_remove:
            del gm_agent.sub_agents[key]

        # Then register new NPC agents for active NPCs
        for npc_id in active_npc_ids:
            npc_data = world_pack.get_npc(npc_id)
            if npc_data:
                # Import LLM from gm_agent
                npc_agent = NPCAgent(llm=gm_agent.llm, vector_store=gm_agent.vector_store)
                agent_key = f"npc_{npc_id}"
                gm_agent.sub_agents[agent_key] = npc_agent

        # Update game state with new session
        gm_agent.game_state.session_id = session_id
        gm_agent.game_state.world_pack_id = request.world_pack_id
        gm_agent.game_state.current_location = starting_location_id
        gm_agent.game_state.active_npc_ids = active_npc_ids

        # Set player (user) name - distinct from character name
        gm_agent.game_state.player_name = request.player_name

        # Handle preset character selection
        if request.preset_character_id:
            preset = world_pack.get_preset_character(request.preset_character_id)
            if preset is None:
                available_presets = [p.id for p in world_pack.preset_characters]
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Preset character not found: {request.preset_character_id}. "
                    f"Available: {available_presets}",
                )
            # Use preset character data
            gm_agent.game_state.player.name = preset.name
            gm_agent.game_state.player.concept = preset.concept
            gm_agent.game_state.player.traits = preset.traits.copy()

        # Reset game state for new session
        gm_agent.game_state.messages = []
        gm_agent.game_state.turn_count = 0
        gm_agent.game_state.last_check_result = None
        gm_agent.game_state.temp_context = {}
        gm_agent.game_state.flags = set()
        gm_agent.game_state.discovered_items = set()

        # Build player data response
        player_data = {
            "name": gm_agent.game_state.player.name,
            "concept": gm_agent.game_state.player.concept.model_dump(),
            "traits": [t.model_dump() for t in gm_agent.game_state.player.traits],
            "tags": gm_agent.game_state.player.tags,
            "fate_points": gm_agent.game_state.player.fate_points,
        }

        # Build game state response
        game_state_data = {
            "current_location": starting_location_id,
            "current_phase": gm_agent.game_state.current_phase.value,
            "turn_count": gm_agent.game_state.turn_count,
            "active_npc_ids": active_npc_ids,
        }

        # Build world info response
        world_info = {
            "id": request.world_pack_id,
            "name": world_pack.info.name.model_dump(),
            "description": world_pack.info.description.model_dump(),
            "version": world_pack.info.version,
            "author": world_pack.info.author,
        }

        # Add optional setting and player_hook if available
        info_dict = world_pack.info.model_dump() if hasattr(world_pack.info, "model_dump") else {}
        if "setting" in info_dict:
            world_info["setting"] = info_dict["setting"]
        if "player_hook" in info_dict:
            world_info["player_hook"] = info_dict["player_hook"]

        # Build starting scene response
        starting_scene = {
            "location_id": starting_location_id,
            "location_name": starting_location.name.model_dump(),
            "description": starting_location.description.model_dump(),
            "items": starting_location.items or [],
            "connected_locations": [],
            "npcs": [],
        }

        # Add atmosphere if available
        loc_dict = (
            starting_location.model_dump() if hasattr(starting_location, "model_dump") else {}
        )
        if "atmosphere" in loc_dict and loc_dict["atmosphere"]:
            starting_scene["atmosphere"] = loc_dict["atmosphere"]

        # Add connected location names
        for loc_id in starting_location.connected_locations or []:
            connected_loc = world_pack.get_location(loc_id)
            if connected_loc:
                starting_scene["connected_locations"].append(
                    {
                        "id": loc_id,
                        "name": connected_loc.name.model_dump(),
                    }
                )

        # Add NPC info - only appearance, NOT name (to prevent metagaming)
        for npc_id in active_npc_ids:
            npc = world_pack.get_npc(npc_id)
            if npc:
                npc_info = {
                    "id": npc_id,
                    # Don't include name - player hasn't learned it yet
                }
                # Use appearance if available, otherwise use first sentence of description
                soul_dict = npc.soul.model_dump() if hasattr(npc.soul, "model_dump") else {}
                if "appearance" in soul_dict and soul_dict["appearance"]:
                    npc_info["appearance"] = soul_dict["appearance"]
                else:
                    # Fallback: use first sentence of description (external appearance only)
                    npc_info["appearance"] = npc.soul.description.model_dump()
                starting_scene["npcs"].append(npc_info)

        return NewGameResponse(
            session_id=session_id,
            player=player_data,
            game_state=game_state_data,
            world_info=world_info,
            starting_scene=starting_scene,
            message="Game session created successfully",
        )

    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create game session: {str(exc)}",
        ) from exc


@router.post("/game/action")
async def process_player_action(
    action_data: dict[str, Any],
):
    """
    Process a player action through the GM Agent.

    Args:
        action_data: Dictionary containing:
            - player_input: Player's action/description
            - lang: Language code (cn/en, default: cn)

    Returns:
        Response from GM Agent including any agent calls and narrative
    """
    from src.backend.main import gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    try:
        # Extract player input and language
        player_input = action_data.get("player_input", "")
        lang = action_data.get("lang", "cn")

        if not player_input:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="player_input is required",
            )

        # Process through GM Agent
        result = await gm_agent.process(
            {
                "player_input": player_input,
                "lang": lang,
            }
        )

        # Return response
        return {
            "success": result.success,
            "content": result.content,
            "metadata": result.metadata,
            "error": result.error,
        }

    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to process action: {str(exc)}",
        ) from exc


@router.get("/game/state")
async def get_game_state():
    """
    Get the current game state.

    Returns:
        Current game state including player info, location, scene details, etc.
    """
    from src.backend.main import get_world_pack_loader, gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    try:
        state = gm_agent.game_state
        world_pack_loader = get_world_pack_loader()

        # Basic state response
        response = {
            "session_id": state.session_id,
            "world_pack_id": state.world_pack_id,
            "player_name": state.player_name,  # PL name (user)
            "player": {
                "name": state.player.name,  # PC name (character)
                "concept": state.player.concept.model_dump(),
                "traits": [t.model_dump() for t in state.player.traits],
                "tags": state.player.tags,
                "fate_points": state.player.fate_points,
            },
            "current_location": state.current_location,
            "active_npc_ids": state.active_npc_ids,
            "current_phase": state.current_phase.value,
            "turn_count": state.turn_count,
            "language": state.language,
            "messages": state.messages[-10:],  # Last 10 messages
        }

        # Try to add scene details from world pack
        if world_pack_loader:
            try:
                world_pack = world_pack_loader.load(state.world_pack_id)
                location = world_pack.get_location(state.current_location)
                if location:
                    response["current_scene"] = {
                        "location_name": location.name.model_dump(),
                        "description": location.description.model_dump(),
                        "items": location.items or [],
                        "connected_locations": location.connected_locations or [],
                    }

                    # Add NPC names
                    npcs = []
                    for npc_id in state.active_npc_ids:
                        npc = world_pack.get_npc(npc_id)
                        if npc:
                            npcs.append({"id": npc_id, "name": npc.soul.name})
                    response["current_scene"]["npcs"] = npcs
            except Exception:
                # If world pack loading fails, continue without scene details
                pass

        return response

    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get game state: {str(exc)}",
        ) from exc


@router.post("/game/dice-result")
async def submit_dice_result(
    dice_data: dict[str, Any],
):
    """
    Submit a dice roll result.

    Args:
        dice_data: Dictionary containing:
            - total: Dice roll total
            - all_rolls: List of all dice rolled
            - kept_rolls: List of kept dice
            - outcome: Roll outcome (critical/success/partial/failure)

    Returns:
        Confirmation and next steps
    """
    from src.backend.main import gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    try:
        # Validate required fields
        required_fields = ["total", "all_rolls", "kept_rolls", "outcome"]
        for field in required_fields:
            if field not in dice_data:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Missing required field: {field}",
                )

        # Store dice result in game state
        gm_agent.game_state.last_check_result = dice_data

        # TODO: Process the result through appropriate agents
        # For now, just acknowledge receipt

        return {
            "success": True,
            "message": "Dice result recorded",
            "next_phase": "narrating",
        }

    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to process dice result: {str(exc)}",
        ) from exc


@router.get("/game/messages")
async def get_recent_messages(count: int = 10):
    """
    Get recent game messages.

    Args:
        count: Number of recent messages to retrieve (default: 10)

    Returns:
        List of recent messages
    """
    from src.backend.main import gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    try:
        messages = gm_agent.game_state.get_recent_messages(count=count)
        return {
            "messages": messages,
            "count": len(messages),
        }

    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get messages: {str(exc)}",
        ) from exc


@router.post("/game/reset")
async def reset_game_state():
    """
    Reset the game to initial state.

    Returns:
        Confirmation of reset
    """
    from src.backend.main import gm_agent

    if gm_agent is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Game engine not initialized",
        )

    try:
        # Reset game state
        gm_agent.game_state.messages = []
        gm_agent.game_state.turn_count = 0
        gm_agent.game_state.last_check_result = None
        gm_agent.game_state.temp_context = {}
        gm_agent.game_state.flags = set()
        gm_agent.game_state.discovered_items = set()
        gm_agent.game_state.set_phase("waiting_input")

        return {
            "success": True,
            "message": "Game state reset",
        }

    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to reset game: {str(exc)}",
        ) from exc


@router.get("/game/world-packs")
async def list_world_packs():
    """
    List available world packs.

    Returns:
        List of available world pack IDs
    """
    from src.backend.main import get_world_pack_loader

    world_pack_loader = get_world_pack_loader()
    if world_pack_loader is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="World pack loader not initialized",
        )

    try:
        available = world_pack_loader.list_available()
        return {
            "world_packs": available,
            "count": len(available),
        }
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list world packs: {str(exc)}",
        ) from exc


@router.get("/game/world-pack/{pack_id}")
async def get_world_pack_info(pack_id: str):
    """
    Get information about a specific world pack.

    Args:
        pack_id: World pack identifier

    Returns:
        World pack metadata and summary
    """
    from src.backend.main import get_world_pack_loader

    world_pack_loader = get_world_pack_loader()
    if world_pack_loader is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="World pack loader not initialized",
        )

    try:
        world_pack = world_pack_loader.load(pack_id)

        # Build info dict with optional setting and player_hook
        info_dict = {
            "name": world_pack.info.name.model_dump(),
            "description": world_pack.info.description.model_dump(),
            "version": world_pack.info.version,
            "author": world_pack.info.author,
        }

        # Add setting if available
        if world_pack.info.setting:
            info_dict["setting"] = {
                "era": world_pack.info.setting.era.model_dump(),
                "genre": world_pack.info.setting.genre.model_dump(),
                "tone": world_pack.info.setting.tone.model_dump(),
            }

        # Add player_hook if available
        if world_pack.info.player_hook:
            info_dict["player_hook"] = world_pack.info.player_hook.model_dump()

        return {
            "id": pack_id,
            "info": info_dict,
            "summary": {
                "locations": len(world_pack.locations),
                "npcs": len(world_pack.npcs),
                "lore_entries": len(world_pack.entries),
                "preset_characters": len(world_pack.preset_characters),
            },
            "locations": [
                {
                    "id": loc_id,
                    "name": loc.name.model_dump(),
                    "tags": loc.tags,
                }
                for loc_id, loc in world_pack.locations.items()
            ],
            "npcs": [
                {
                    "id": npc_id,
                    "name": npc.soul.name,
                    "location": npc.body.location,
                }
                for npc_id, npc in world_pack.npcs.items()
            ],
            "preset_characters": [
                {
                    "id": preset.id,
                    "name": preset.name,
                    "concept": preset.concept.model_dump(),
                    "traits": [t.model_dump() for t in preset.traits],
                }
                for preset in world_pack.preset_characters
            ],
        }

    except FileNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"World pack not found: {pack_id}",
        ) from exc
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get world pack info: {str(exc)}",
        ) from exc

File: src/backend-ts/src/api/v1/game.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import { v4 as uuidv4 } from "uuid";
import { getAppContext } from "../../index";
import type { GameState, PlayerCharacter, Trait } from "../../schemas";
import { GMAgent } from "../../agents/gm";
import { NPCAgent } from "../../agents/npc";
import { ConfigService } from "../../services/config";
import { LLMFactory } from "../../lib/llm-factory";

const NewGameRequestSchema = z.object({
  world_pack_id: z.string().default("demo_pack"),
  player_name: z.string().default("玩家"),
  preset_character_id: z.string().optional(),
});

const PlayerActionRequestSchema = z.object({
  session_id: z.string(),
  action: z.string(),
  lang: z.enum(["cn", "en"]).default("cn"),
});

const DiceResultRequestSchema = z.object({
  session_id: z.string(),
  rolls: z.array(z.number()),
  total: z.number(),
  outcome: z.string(),
  modifier: z.number().default(0),
});

export const gameRouter = new Hono();

gameRouter.post(
  "/game/new",
  zValidator("json", NewGameRequestSchema),
  async (c) => {
    const request = c.req.valid("json");
    const ctx = getAppContext();

    if (!ctx.worldPackLoader) {
      return c.json(
        { error: "World pack loader not initialized" },
        503
      );
    }

    try {
      const sessionId = uuidv4();

      let worldPack;
      try {
        worldPack = await ctx.worldPackLoader.load(request.world_pack_id);
      } catch (error) {
        const available = await ctx.worldPackLoader.listAvailable();
        return c.json(
          {
            error: `World pack not found: ${request.world_pack_id}. Available: ${available.join(", ")}`,
          },
          404
        );
      }

      let startingLocationId: string | null = null;
      let startingLocation = null;

      for (const [locId, loc] of Object.entries(worldPack.locations)) {
        if (loc.tags?.includes("starting_area")) {
          startingLocationId = locId;
          startingLocation = loc;
          break;
        }
      }

      if (!startingLocationId && Object.keys(worldPack.locations).length > 0) {
        startingLocationId = Object.keys(worldPack.locations)[0] || null;
        startingLocation = startingLocationId ? worldPack.locations[startingLocationId] : null;
      }

      if (!startingLocation) {
        return c.json({ error: "World pack has no locations defined" }, 400);
      }

      const activeNpcIds = startingLocation.present_npc_ids || [];
      const config = ConfigService.getInstance().get();
      
      // Initialize Agents - Require LLM configuration
      if (!config.agents || !config.providers || config.providers.length === 0) {
        return c.json(
          { error: "Game engine not initialized. Configure LLM settings first." },
          503
        );
      }

      let gmModel, npcModel;
      try {
        gmModel = LLMFactory.createModel(config.agents.gm, config.providers);
        npcModel = LLMFactory.createModel(config.agents.npc, config.providers);
        console.log("🤖 Agents initialized with real LLM configuration");
      } catch (err) {
        console.error("❌ Failed to create LLM models:", err);
        return c.json(
          { error: `Failed to initialize AI models: ${err}` },
          503
        );
      }

      let playerCharacter: PlayerCharacter;

      if (request.preset_character_id) {
        const preset = worldPack.preset_characters?.find(
          (p: any) => p.id === request.preset_character_id
        );
        if (!preset) {
          const availableIds = worldPack.preset_characters?.map((p: any) => p.id) || [];
          return c.json(
            {
              error: `Preset character not found: ${request.preset_character_id}. Available: ${availableIds.join(", ")}`,
            },
            400
          );
        }
        playerCharacter = {
          name: preset.name,
          concept: preset.concept,
          traits: preset.traits,
          fate_points: 3,
          tags: [],
        };
      } else {
        const defaultTrait: Trait = {
          name: { cn: "勇敢", en: "Brave" },
          description: {
            cn: "面对困难不退缩",
            en: "Faces difficulties without retreat",
          },
          positive_aspect: { cn: "勇敢", en: "Brave" },
          negative_aspect: { cn: "鲁莽", en: "Rash" },
        };

        playerCharacter = {
          name: request.player_name || "冒险者",
          concept: { cn: "冒险者", en: "Adventurer" },
          traits: [defaultTrait],
          fate_points: 3,
          tags: [],
        };
      }

      const gameState: GameState = {
        session_id: sessionId,
        player_name: request.player_name || "玩家",
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
        player: playerCharacter,
        current_phase: "waiting_input",
        next_agent: null,
        world_pack_id: request.world_pack_id,
        current_location: startingLocationId || "",
        active_npc_ids: activeNpcIds,
        discovered_items: [],
        flags: [],
        game_time: "00:00",
        turn_count: 0,
        messages: [],
        temp_context: {},
        last_check_result: null,
        react_pending_state: null,
        language: "cn",
      };

      const worldInfo: any = {
        id: request.world_pack_id,
        name: worldPack.info.name,
        description: worldPack.info.description,
        version: worldPack.info.version,
        author: worldPack.info.author,
      };

      if (worldPack.info.setting) {
        worldInfo.setting = worldPack.info.setting;
      }
      if (worldPack.info.player_hook) {
        worldInfo.player_hook = worldPack.info.player_hook;
      }

      const startingScene = {
        location_id: startingLocationId,
        location_name: startingLocation.name,
        description: startingLocation.description, // Aligned with Python backend
        items: startingLocation.items || [],
        connected_locations: [], // Populated below
        npcs: [], // Populated below
        atmosphere: startingLocation.atmosphere, // Add atmosphere if present
      };

      // Populate connected locations
      if (startingLocation.connected_locations) {
        for (const locId of startingLocation.connected_locations) {
          const connectedLoc = worldPack.locations[locId];
          if (connectedLoc) {
            (startingScene.connected_locations as any[]).push({
              id: locId,
              name: connectedLoc.name, // Pass raw LocalizedString object
            });
          }
        }
      }

      // Populate NPCs - Hide names to prevent metagaming
      for (const npcId of activeNpcIds) {
        const npc = Object.values(worldPack.npcs || {}).find((n: any) => n.id === npcId);
        if (npc) {
            const npcInfo: any = { id: npcId };
            // Use appearance if available, otherwise fallback to description
            const soul = (npc as any).soul;
            if (soul.appearance) {
                npcInfo.appearance = soul.appearance;
            } else {
                // Fallback: use description (Python sends full description model_dump)
                npcInfo.appearance = soul.description;
            }
            (startingScene.npcs as any[]).push(npcInfo);
        }
      }

      // NO message push to gameState.messages - Python backend starts empty

      const npcMaxTokens = config.agents?.npc?.max_tokens;
      const gmMaxTokens = config.agents?.gm?.max_tokens;

      const npcAgent = new NPCAgent(npcModel as any, npcMaxTokens);
      const subAgents = {
        npc: npcAgent,
      };

      ctx.gmAgent = new GMAgent(
        gmModel as any,
        subAgents,
        gameState,
        ctx.loreService,
        ctx.worldPackLoader,
        gmMaxTokens
      );

      return c.json({
        session_id: sessionId,
        player: playerCharacter,
        game_state: {
            ...gameState,
            current_phase: gameState.current_phase,
            turn_count: gameState.turn_count,
            active_npc_ids: activeNpcIds,
        },
        world_info: worldInfo,
        starting_scene: startingScene,
        message: "Game session created successfully",
      });
    } catch (error) {
      console.error("Error starting new game:", error);
      return c.json(
        { error: `Failed to start game: ${error}` },
        500
      );
    }
  }
);

gameRouter.post(
  "/game/action",
  zValidator("json", PlayerActionRequestSchema),
  async (c) => {
    const request = c.req.valid("json");
    const ctx = getAppContext();

    if (!ctx.gmAgent) {
      return c.json(
        {
          error: "Game engine not initialized. Configure LLM settings first.",
        },
        503
      );
    }

    try {
      console.log(`[GameAPI] Received action: ${request.action}`);
      const response = await ctx.gmAgent.process({
        player_input: request.action,
        lang: request.lang,
      });
      console.log(`[GameAPI] Process finished. Success: ${response.success}`);

      return c.json({
        success: response.success,
        content: response.content,
        metadata: response.metadata,
        error: response.error,
      });
    } catch (error) {
      console.error("Error processing action:", error);
      return c.json(
        { error: `Failed to process action: ${error}` },
        500
      );
    }
  }
);

gameRouter.post(
  "/game/dice-result",
  zValidator("json", DiceResultRequestSchema),
  async (c) => {
    const request = c.req.valid("json");
    const ctx = getAppContext();

    if (!ctx.gmAgent) {
      return c.json(
        { error: "Game engine not initialized" },
        503
      );
    }

    try {
      const response = await ctx.gmAgent.resumeAfterDice(
        {
          rolls: request.rolls,
          total: request.total,
          outcome: request.outcome,
          modifier: request.modifier,
        },
        "cn"
      );

      return c.json({
        success: response.success,
        content: response.content,
        metadata: response.metadata,
        error: response.error,
      });
    } catch (error) {
      console.error("Error processing dice result:", error);
      return c.json(
        { error: `Failed to process dice result: ${error}` },
        500
      );
    }
  }
);

gameRouter.get("/game/state", async (c) => {
  const ctx = getAppContext();

  if (!ctx.gmAgent) {
    return c.json({ error: "Game engine not initialized" }, 503);
  }

  const gameState = ctx.gmAgent.getGameState();

  return c.json({
    session_id: gameState.session_id,
    current_phase: gameState.current_phase, // Aligned with Python backend (was 'phase')
    turn_number: gameState.turn_count,
    current_location: gameState.current_location,
    active_npc_ids: gameState.active_npc_ids,
    player: gameState.player,
  });
});

gameRouter.get("/game/state/:sessionId", async (c) => {
  const sessionId = c.req.param("sessionId");
  const ctx = getAppContext();

  if (!ctx.gmAgent) {
    return c.json({ error: "Game engine not initialized" }, 503);
  }

  const gameState = ctx.gmAgent.getGameState();

  if (gameState.session_id !== sessionId) {
    return c.json({ error: "Session not found" }, 404);
  }

  return c.json({
    session_id: gameState.session_id,
    current_phase: gameState.current_phase, // Aligned with Python backend (was 'phase')
    turn_number: gameState.turn_count,
    current_location: gameState.current_location,
    active_npc_ids: gameState.active_npc_ids,
    player: gameState.player,
  });
});

  gameRouter.get("/game/world-packs", async (c) => {
  const ctx = getAppContext();

  if (!ctx.worldPackLoader) {
    return c.json({ error: "World pack loader not initialized" }, 503);
  }

  try {
    const available = await ctx.worldPackLoader.listAvailable();
    return c.json({ packs: available });
  } catch (error) {
    return c.json({ error: `Failed to list world packs: ${error}` }, 500);
  }
});

  gameRouter.get("/game/world-pack/:packId", async (c) => {
  const packId = c.req.param("packId");
  const ctx = getAppContext();

  if (!ctx.worldPackLoader) {
    return c.json({ error: "World pack loader not initialized" }, 503);
  }

  try {
    const worldPack = await ctx.worldPackLoader.load(packId);

    const locations = Object.entries(worldPack.locations).map(([id, location]) => ({
      id,
      name: location.name,
      tags: location.tags || [],
    }));

    const npcs = Object.entries(worldPack.npcs || {}).map(([id, npc]) => ({
      id,
      name: npc.soul.name,
      location: npc.body?.location || "unknown",
    }));

    return c.json({
      id: packId,
      info: {
        name: worldPack.info.name,
        description: worldPack.info.description,
        version: worldPack.info.version,
        author: worldPack.info.author,
        setting: worldPack.info.setting,
        player_hook: worldPack.info.player_hook,
      },
      summary: {
        locations: Object.keys(worldPack.locations).length,
        npcs: Object.keys(worldPack.npcs || {}).length,
        lore_entries: Object.keys(worldPack.entries).length,
        preset_characters: worldPack.preset_characters?.length || 0,
      },
      locations,
      npcs,
      preset_characters: worldPack.preset_characters || [],
    });
  } catch (error) {
    return c.json({ error: `World pack not found: ${packId}` }, 404);
  }
});
